src/test/java/org/opensearch/security/support/SecurityIndexHandlerTest.java

```diff
@@ -0,0 +1,384 @@
+        when(client.threadPool()).thenReturn(threadPool);
+        threadContext = new ThreadContext(Settings.EMPTY);
+        when(client.threadPool()).thenReturn(threadPool);
+        when(threadPool.getThreadContext()).thenReturn(threadContext);
+        configFolder = temporaryFolder.newFolder("config").toPath();
+        securityIndexHandler = new SecurityIndexHandler(INDEX_NAME, Settings.EMPTY, client);
+    }
+
+    @Test
+    public void createIndex_shouldCreateIndex() {
+        doAnswer(invocation -> {
+            ActionListener<CreateIndexResponse> actionListener = invocation.getArgument(1);
+            actionListener.onResponse(new CreateIndexResponse(true, true, "some_index"));
+            return null;
+        }).when(indicesAdminClient).create(any(), any());
+
+        securityIndexHandler.createIndex(ActionListener.wrap(Assert::assertTrue, Assert::assertNull));
+
+        final var requestCaptor = ArgumentCaptor.forClass(CreateIndexRequest.class);
+
+        verify(indicesAdminClient).create(requestCaptor.capture(), any());
+
+        final var createRequest = requestCaptor.getValue();
+        assertEquals(INDEX_NAME, createRequest.index());
+        for (final var setting : SecurityIndexHandler.INDEX_SETTINGS.entrySet())
+            assertEquals(setting.getValue().toString(), createRequest.settings().get(setting.getKey()));
+
+        assertEquals(ActiveShardCount.ONE, createRequest.waitForActiveShards());
+    }
+
+    @Test
+    public void createIndex_shouldReturnSecurityExceptionIfItCanNotCreateIndex() {
+
+        final var listener = spy(ActionListener.<Boolean>wrap(r -> fail("Unexpected behave"), e -> {
+            assertEquals(SecurityException.class, e.getClass());
+            assertEquals("Couldn't create security index " + INDEX_NAME, e.getMessage());
+        }));
+
+        doAnswer(invocation -> {
+            ActionListener<CreateIndexResponse> actionListener = invocation.getArgument(1);
+            actionListener.onResponse(new CreateIndexResponse(false, false, "some_index"));
+            return null;
+        }).when(indicesAdminClient).create(any(), any());
+
+        securityIndexHandler.createIndex(listener);
+
+        verify(indicesAdminClient).create(isA(CreateIndexRequest.class), any());
+        verify(listener).onFailure(any());
+    }
+
+    @Test
+    public void uploadDefaultConfiguration_shouldFailIfRequiredConfigFilesAreMissing() {
+        final var listener = spy(ActionListener.<Set<SecurityConfig>>wrap(r -> fail("Unexpected behave"), e -> {
+            assertEquals(SecurityException.class, e.getClass());
+            assertThat(e.getMessage(), containsString("Couldn't find configuration file"));
+        }));
+        securityIndexHandler.uploadDefaultConfiguration(configFolder, listener);
+
+        verify(listener).onFailure(any());
+    }
+
+    @Test
+    public void uploadDefaultConfiguration_shouldFailIfBulkHasFailures() throws IOException {
```

Comment:
is the bulk transaction atomic? or is partial success allowed ?

Reply:
OS does not support transactions. If bulk insertion fails the cluster state never updates state as result no cache reload on nodes.

Reply:
So no half updates would be allowed. Meaning `roles` upload didn't fail but `rolesmapping` did, then both with fail? essentially, all or nothing?

Reply:
Half update is not allowed. It depends what the end user wanna to upload. The current solution is to check config.yml file only while the new one checks the list of important files the importance of the file is in the `CType` enum aka `allowEpmtyFile`